core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java

```diff
@@ -60,28 +74,50 @@ public Object getFieldValue(Metric o) {
   private IndexedSet<Metric> getMetricsByInstanceType(MetricsSystem.InstanceType instanceType) {
     if (instanceType == InstanceType.WORKER) {
       return mWorkerMetrics;
+    } else if (instanceType == InstanceType.CLIENT) {
+      return mClientMetrics;
     } else {
       throw new IllegalArgumentException("Unsupported instance type " + instanceType);
     }
   }
 
   /**
-   * Put the metrics from an instance with a hostname. If all the old metrics associated with this
+   * Put the metrics from a worker with a hostname. If all the old metrics associated with this
    * instance will be removed and then replaced by the latest.
    *
-   * @param instance the instance type
    * @param hostname the hostname of the instance
    * @param metrics the new worker metrics
    */
-  public synchronized void putWorkerMetrics(MetricsSystem.InstanceType instance, String hostname,
+  public synchronized void putWorkerMetrics(String hostname, List<Metric> metrics) {
+    if (metrics.isEmpty()) {
+      return;
+    }
+    mWorkerMetrics.removeByField(HOSTNAME_ID_INDEX, hostname);
+    for (Metric metric : metrics) {
+      mWorkerMetrics.add(metric);
+    }
+  }
+
+  /**
+   * Put the metrics from a client with a hostname and a client id. If all the old metrics
+   * associated with this instance will be removed and then replaced by the latest.
+   *
+   * @param hostname the hostname of the client
+   * @param clientId the id of the client
+   * @param metrics the new metrics
+   */
+  public synchronized void putClientMetrics(String hostname, String clientId,
       List<Metric> metrics) {
-    IndexedSet<Metric> set = getMetricsByInstanceType(instance);
-    set.removeByField(HOSTNAME_INDEX, hostname);
+    if (metrics.isEmpty()) {
+      return;
+    }
+    mClientMetrics.removeByField(HOSTNAME_ID_INDEX, getHostnameAndId(hostname, clientId));
     for (Metric metric : metrics) {
       if (metric.getHostname() == null) {
         continue; // ignore metrics who hostname is null
```

Comment:
why would the hostname be null? Is this check needed for `putWorkerMetrics` as well?

Reply:
This is a safe guard. It can happen in our simulated local cluster in integration testing, where the same JVM has the master, worker and client metrics.

Reply:
should we use it in `putWorkerMetrics` then?

Reply:
yup, updated `putWorkerMetrics `